contracts/examples/digital-cash/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+#![no_std]
+#![allow(unused_attributes)]
+
+use elrond_wasm::{imports, require, sc_error};
+imports!();
+
+mod deposit_info;
+use deposit_info::DepositInfo;
+
+pub const SECONDS_PER_ROUND: u64 = 6;
+
+#[elrond_wasm_derive::contract(DepositImpl)]
+pub trait Deposit {
+    #[init]
+    fn init(&self) {}
+
+    fn get_expiration_round(&self, valability: u64) -> u64 {
+        let valability_rounds = valability / SECONDS_PER_ROUND;
+
+        return self.get_block_round() + valability_rounds;
+    }
+
+    #[endpoint]
+    #[payable("*")]
+    fn fund(
+        &self,
+        #[payment] payment: BigUint,
+        #[payment_token] token: TokenIdentifier,
+        address: Address,
+        valability: u64,
+    ) -> SCResult<()> {
+        require!(payment > 0, "amount must be greater than 0");
+        require!(self.deposit(&address).is_empty(), "key already used");
+
+        let deposit = &DepositInfo {
+            amount: payment,
+            depositor_address: self.get_caller(),
+            expiration_round: self.get_expiration_round(valability),
+            token_name: token,
+        };
+
+        self.deposit(&address).set(deposit);
+
+        Ok(())
+    }
+
+    #[endpoint]
+    fn withdraw(&self, address: Address) -> SCResult<()> {
+        let deposit = self.deposit(&address).get();
```

Comment:
On the topic of validations...
The `withdraw`, `claim` and `get_amount` methods all assume that there is a deposit at the given address. However, this is not guaranteed.
Even though the methods probably fail in some way, which is to be expected, there should be a check for this. Otherwise we get cryptic error messages.

Reply:
Done!